lib/flight_scheduler/stepd.rb

```diff
@@ -208,18 +215,39 @@ def create_input_thread(input_wr, connection)
       end
     end
 
-    def notify_started
-      MessageSender.send({
+    def notify_started(connection)
+      connection.write({
         command: 'RUN_STEP_STARTED',
         job_id: @job.id,
         port: @port,
         step_id: @step.id,
       })
+      connection.flush
     end
 
-    def notify_finished(status)
+    def notify_finished(status, connection)
       command = status.success? ? 'RUN_STEP_COMPLETED' : 'RUN_STEP_FAILED'
-      MessageSender.send(command: command, job_id: @job.id, step_id: @step.id)
+      connection.write(command: command, job_id: @job.id, step_id: @step.id)
+      connection.flush
+    end
+
+    def with_controller_connection(&block)
+      controller_url = FlightScheduler.app.config.controller_url
+      endpoint = Async::HTTP::Endpoint.parse(controller_url)
+      auth_token = FlightScheduler::Auth.token
+
+      Async.logger.info("Connecting to #{controller_url.inspect}") { endpoint }
+      Async::WebSocket::Client.connect(endpoint) do |connection|
```

Comment:
Ahhh I see why the pre-existing reactor now causes a problem. I think using `Async` to implement websockets is the underlining problem here.

Using `Async` in the main daemon process makes sense as it avoids spinning up a new control thread for each message it receives. However I don't think this is required in the `stepd` and `batchd`. Possible a different web-socket library is required to handle it synchronously.

Reply:
For future reference, it maybe possible to use `faye` without `EventMachine`:
https://github.com/faye/websocket-driver-ruby